{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 117,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import sys\n",
    "import csv\n",
    "import json\n",
    "\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import scipy.stats as stats\n",
    "\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "from tqdm.notebook import tqdm\n",
    "from collections import defaultdict\n",
    "\n",
    "from scipy.optimize import curve_fit\n",
    "from scipy.special import factorial\n",
    "from scipy.stats import poisson\n",
    "from scipy.stats import nbinom\n",
    "from scipy.stats import chi2\n",
    "\n",
    "%matplotlib inline"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 118,
   "metadata": {},
   "outputs": [],
   "source": [
    "minimum_maf = 0.01"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Load the permuation/randomization study results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 119,
   "metadata": {},
   "outputs": [],
   "source": [
    "# we did an analysis to compare the maf distribution\n",
    "# this makes sure we grab only those that are similar\n",
    "\n",
    "maf_distribution_stats_files = (\n",
    "    ('gs2', 'data/ukbb/gene_set_2/gene_set_2.json'),\n",
    ")\n",
    "\n",
    "run2excluded_samples = dict()\n",
    "\n",
    "for run_code, maf_stats_file in maf_distribution_stats_files:\n",
    "    fh = open(maf_stats_file)\n",
    "    results = json.load(fh)\n",
    "    fh.close()\n",
    "    \n",
    "    excluded_samples = [r['name'].strip('.txt') for r in results['sample_stats'] if r['Up'] < 0.05]\n",
    "    run2excluded_samples[run_code] = excluded_samples\n",
    "    \n",
    "    del excluded_samples\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 130,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "May exclude up to 0 samples.\n",
      " Excluded 0 samples.\n",
      "May exclude up to 0 samples.\n",
      " Excluded 0 samples.\n",
      "May exclude up to 52 samples.\n",
      " Excluded 52 samples.\n",
      "May exclude up to 52 samples.\n",
      " Excluded 52 samples.\n",
      "May exclude up to 52 samples.\n",
      " Excluded 0 samples.\n",
      "May exclude up to 52 samples.\n",
      " Excluded 0 samples.\n",
      "May exclude up to 0 samples.\n",
      " Excluded 0 samples.\n",
      "May exclude up to 0 samples.\n",
      " Excluded 0 samples.\n"
     ]
    }
   ],
   "source": [
    "random_sample_files = (\n",
    "    # gene_set_kegg\n",
    "    ('kegg_apr_snp', 'data/ukbb/gene_set_kegg/ukbb_plinkresults.old_genesnpssamples_1.txt'),\n",
    "    ('kegg_apr_phe', 'data/ukbb/gene_set_kegg/ukbb_phenpermuteresults.txt'),\n",
    "    # Gene Set 2\n",
    "    ('gs2_apr_snp', 'data/ukbb/gene_set_2/ukbb_plinkresults_genesnpssamples_2_pos_severe_apr_bydist_0to80kb.txt'),\n",
    "    ('gs2_may_snp', 'data/ukbb/gene_set_2/ukbb_plinkresults_genesnpssamples_2_pos_severe_may_bydist_0to80kb.txt'),\n",
    "    ('gs2_apr_phe', 'data/ukbb/gene_set_2/ukbb_phenpermuteresults_pos_severe_apr_snps_gene_2_bydist_0to80kb.txt'),\n",
    "    ('gs2_may_phe', 'data/ukbb/gene_set_2/ukbb_phenpermuteresults_pos_severe_may_snps_gene_2_bydist_0to80kb.txt'),\n",
    "    # Gene Set LD\n",
    "    ('ld_apr_phe', 'data/ukbb/gene_set_ld/ukbb_phenpermuteresults_pos_severe_apr_snps_gene_ld_bydist_0to80kb.txt'),\n",
    "    ('ld_may_phe', 'data/ukbb/gene_set_ld/ukbb_phenpermuteresults_pos_severe_may_snps_gene_ld_bydist_0to80kb.txt')\n",
    ")\n",
    "\n",
    "run2dist2alpha2sig = dict()\n",
    "run2output_file_prefix = dict()\n",
    "\n",
    "for run_code, random_sample_file in random_sample_files:\n",
    "\n",
    "    run2output_file_prefix[run_code] = random_sample_file.split('/')[-1].split('.')[0]\n",
    "    \n",
    "    excluded_samples = run2excluded_samples.get(run_code.split('_')[0], list())\n",
    "    print(\"May exclude up to %d samples.\" % len(excluded_samples))\n",
    "    \n",
    "    fh = open(random_sample_file)\n",
    "    reader = csv.reader(fh, delimiter='\\t')\n",
    "    header = next(reader)\n",
    "\n",
    "    dist2alpha2sig = defaultdict(lambda: defaultdict(list))\n",
    "    alphas = set()\n",
    "    \n",
    "    excluded_set = set()\n",
    "    for row in reader:\n",
    "        if len(row) == 6:\n",
    "            sid, maf, dist, alpha, numvar, sigvar = row\n",
    "        elif len(row) == 4:\n",
    "            sid, alpha, numvar, sigvar = row\n",
    "            dist = 60000\n",
    "            maf = 0.001\n",
    "        else:\n",
    "            raise Exception(\"Row is unexpected length.\")\n",
    "        \n",
    "        if float(maf) != minimum_maf:\n",
    "            continue\n",
    "        \n",
    "        if sid in excluded_samples:\n",
    "            excluded_set.add(sid)\n",
    "            continue\n",
    "        \n",
    "        dist2alpha2sig[int(dist)][float(alpha)].append(int(sigvar))\n",
    "        alphas.add(float(alpha))\n",
    "    \n",
    "    alphas = sorted(alphas)\n",
    "    run2dist2alpha2sig[run_code] = dist2alpha2sig\n",
    "    \n",
    "    print(\" Excluded %d samples.\" % len(excluded_set))\n",
    "    del dist2alpha2sig\n",
    "    fh.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Load the Association Study results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 131,
   "metadata": {},
   "outputs": [],
   "source": [
    "plink_results_files = (\n",
    "    ('gs2_apr', 'data/ukbb/gene_set_2/apr_pos_severe.PHENO1.glm.logistic.adjusted'),\n",
    "    ('gs2_may', 'data/ukbb/gene_set_2/may_pos_severe.PHENO1.glm.logistic.adjusted'),\n",
    "    ('ld_apr', 'data/ukbb/gene_set_ld/apr_pos_severe.PHENO1.glm.logistic.adjusted'),\n",
    "    ('ld_may', 'data/ukbb/gene_set_ld/may_pos_severe.PHENO1.glm.logistic.adjusted')\n",
    ")\n",
    "\n",
    "run2plink_results = dict()\n",
    "\n",
    "for run_code, plink_results_file in plink_results_files:\n",
    "\n",
    "    fh = open(plink_results_file)\n",
    "    reader = csv.reader(fh, delimiter='\\t')\n",
    "\n",
    "    header = next(reader)\n",
    "    \n",
    "    plink_results = list()\n",
    "\n",
    "    for row in reader:\n",
    "        if row[0][0] == '#':\n",
    "            continue\n",
    "\n",
    "        chrom = row[0]\n",
    "        rsid = row[1]\n",
    "        allele = row[2]\n",
    "        unadjusted = float(row[3])\n",
    "        fdrbh = float(row[9])\n",
    "\n",
    "        if True or snp_allele_counts[rsid]['maf'] > minimum_maf:\n",
    "            plink_results.append([chrom, rsid, allele, unadjusted, fdrbh])\n",
    "        \n",
    "    run2plink_results[run_code] = plink_results\n",
    "    \n",
    "    del plink_results\n",
    "    fh.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 132,
   "metadata": {},
   "outputs": [],
   "source": [
    "# load the distances for the actual results\n",
    "plink_distance_files = (\n",
    "    ('gs2', 'data/ukbb/gene_set_2/05172020_snps_gene_distance.txt'),\n",
    "    ('ld', 'data/ukbb/gene_set_ld/05242020_snps_gene_distance.txt')\n",
    ")\n",
    "\n",
    "run2plink_rsid2distance = dict()\n",
    "\n",
    "for run_code, plink_results_distance_file in plink_distance_files:\n",
    "    \n",
    "    fh = open(plink_results_distance_file)\n",
    "    reader = csv.reader(fh, delimiter='\\t')\n",
    "\n",
    "    header = next(reader)\n",
    "\n",
    "    plink_rsid2distance = dict()\n",
    "    for rsid, closest_gene, distance in reader:\n",
    "        plink_rsid2distance[rsid] = int(distance)\n",
    "    \n",
    "    run2plink_rsid2distance[run_code] = plink_rsid2distance\n",
    "    \n",
    "    del plink_rsid2distance\n",
    "    fh.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Evaluate the results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 133,
   "metadata": {},
   "outputs": [],
   "source": [
    "# dist = 40000\n",
    "\n",
    "# pvalues = [row[3] for row in plink_results if plink_rsid2distance[row[1]] < dist]\n",
    "\n",
    "# x, y, ylo, yhi = zip(*[(a, np.percentile(s, 50), np.percentile(s, 50)-np.percentile(s, 2.5), np.percentile(s, 97.5)-np.percentile(s, 50)) for a, s in dist2alpha2sig[dist].items()])\n",
    "# plt.errorbar(x, y, yerr=(ylo, yhi), fmt='ko', lw=1)\n",
    "\n",
    "# y_act = [len([p for p in pvalues if p <= a]) for a in x]\n",
    "# plt.plot(x, y_act, 'rd')\n",
    "\n",
    "\n",
    "# plt.xscale('log')\n",
    "# plt.yscale('log')\n",
    "# sns.despine(trim=False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 134,
   "metadata": {},
   "outputs": [],
   "source": [
    "# run this the first time only, the next part can take a while and this will keep the results in memory\n",
    "dist2run2results = dict()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 135,
   "metadata": {},
   "outputs": [],
   "source": [
    "distribution_functions = [\n",
    "    {\n",
    "        'name': 'chi-squared',\n",
    "        'pdf': chi2.pdf,\n",
    "        'cdf': chi2.cdf,\n",
    "        'mean': chi2.mean,\n",
    "        'nparams': 1,\n",
    "        'p0': [2,],\n",
    "    },\n",
    "    {\n",
    "        'name': 'negative-binomial',\n",
    "        'pdf': nbinom.pmf,\n",
    "        'cdf': nbinom.cdf,\n",
    "        'mean': nbinom.mean,\n",
    "        'nparams': 2,\n",
    "        'p0': [1, 0.01]\n",
    "    },\n",
    "    {\n",
    "        'name': 'poisson',\n",
    "        'pdf': poisson.pmf,\n",
    "        'cdf': poisson.cdf,\n",
    "        'mean': poisson.mean,\n",
    "        'nparams': 1,\n",
    "        'p0': [0.3,],\n",
    "    }\n",
    "]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "for dist_func in distribution_functions:\n",
    "    \n",
    "    #print(dist_func)\n",
    "    \n",
    "    fit_function = dist_func['pdf']\n",
    "    p0 = dist_func['p0']\n",
    "    cdf = dist_func['cdf']\n",
    "    \n",
    "    run2results = dist2run2results.get(dist_func['name'], dict())\n",
    "    \n",
    "    for plink_run_code in run2plink_results.keys():\n",
    "        \n",
    "        for perm_method in ('snp', 'phe'):\n",
    "\n",
    "            random_run_code = \"%s_%s\" % (plink_run_code, perm_method)\n",
    "            print(dist_func['name'], random_run_code)\n",
    "            \n",
    "            # check if we already computed the results for this combo\n",
    "            if not random_run_code.startswith('ld') and random_run_code in run2results:\n",
    "                continue\n",
    "            \n",
    "            # make sure we have the data we need to run this combo\n",
    "            if not random_run_code in run2dist2alpha2sig:\n",
    "                continue\n",
    "            \n",
    "            dist_run_code = plink_run_code.split('_')[0]\n",
    "\n",
    "            #print(plink_run_code, random_run_code, dist_run_code)\n",
    "\n",
    "            actual_pvalues = [(row[3], run2plink_rsid2distance[dist_run_code][row[1]]) for row in run2plink_results[plink_run_code]]\n",
    "\n",
    "            results = list()\n",
    "            for dist, alpha2sig in tqdm(run2dist2alpha2sig[random_run_code].items()):\n",
    "\n",
    "                for alpha, sigcs in alpha2sig.items():\n",
    "                    #print(alpha)\n",
    "\n",
    "                    if max(sigcs) <= 1:\n",
    "                        continue\n",
    "\n",
    "                    # how many actual hits are there at this alpha?\n",
    "                    nhits = len([p for p,d in actual_pvalues if p <= alpha and d <= dist])\n",
    "\n",
    "                    plt.figure(figsize=(2,2))\n",
    "                    # the bins should be of integer width, because neg bin is an integer distribution\n",
    "                    bins = np.arange(max(sigcs)+2) - 0.5\n",
    "                    entries, bin_edges, patches = plt.hist(sigcs, bins=bins, density=True, label='Random', color='k')\n",
    "\n",
    "                    # calculate bin centres\n",
    "                    bin_middles = 0.5 * (bin_edges[1:] + bin_edges[:-1])\n",
    "\n",
    "                    # fit with curve_fit\n",
    "                    parameters, cov_matrix = curve_fit(fit_function, bin_middles, entries, p0=p0, maxfev = 6000)\n",
    "                    #print(parameters)\n",
    "\n",
    "                    # plot with fitted parameters\n",
    "                    x_plot = np.arange(0, max(max(sigcs)+2,nhits))\n",
    "\n",
    "                    plt.plot(\n",
    "                        x_plot,\n",
    "                        fit_function(x_plot, *parameters),\n",
    "                        marker='o', linestyle='',\n",
    "                        color='r',\n",
    "                        label='Neg Bin Fit',\n",
    "                    )\n",
    "                    plt.ylabel('Density')\n",
    "                    plt.xlabel('# Rejected H0')\n",
    "                    plt.title(r'$\\alpha = %.2e$' % alpha)\n",
    "\n",
    "                    sigcounts = dict()\n",
    "                    for c in set(sigcs):\n",
    "                        sigcounts[c] = len([x for x in sigcs if x == c])\n",
    "\n",
    "                    actual_counts = [sigcounts[c] for c in sorted(sigcounts)]\n",
    "                    expected_counts = len(sigcs)*fit_function(list(map(int,sorted(sigcounts))), *(parameters))\n",
    "\n",
    "                    if min(expected_counts) == 0:\n",
    "                        expected_counts += 1\n",
    "\n",
    "                    chi, chipvalue = stats.chisquare(\n",
    "                        actual_counts, \n",
    "                        expected_counts\n",
    "                    )\n",
    "\n",
    "                    # p-value for the actual number of hits\n",
    "                    nhits_pvalue = 1-cdf(nhits, *parameters)\n",
    "\n",
    "                    plt.plot([nhits, nhits], [0, max(entries)], 'r--')\n",
    "                    sns.despine(trim=True)\n",
    "\n",
    "                    plt.close()\n",
    "                    results.append({\n",
    "                        'distance': dist, \n",
    "                        'alpha': alpha, \n",
    "                        'num_reps': len(sigcs), \n",
    "                        'max_rep_hits': max(sigcs), \n",
    "                        'fit_p': list(parameters),\n",
    "                        'chisquared': chi, \n",
    "                        'chi_pvalue': chipvalue, \n",
    "                        'nhits': nhits, \n",
    "                        'nhits_pvalue': nhits_pvalue,\n",
    "                        'actual_counts': actual_counts,\n",
    "                        'expected_counts': list(expected_counts)})\n",
    "\n",
    "            run2results[random_run_code] = results\n",
    "    \n",
    "    dist2run2results[dist_func['name']] = run2results\n",
    "    del run2results\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 137,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "dict_keys(['chi-squared', 'negative-binomial', 'poisson'])"
      ]
     },
     "execution_count": 137,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dist2run2results.keys()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 138,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save dist2run2results for later plotting\n",
    "fh = open('results/null_distirbution_fit_results_minmaf%.3f_0to80kb.json' % minimum_maf, 'w')\n",
    "fh.write(json.dumps(dist2run2results))\n",
    "fh.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 150,
   "metadata": {},
   "outputs": [],
   "source": [
    "# determine which alphas to plot automatically from the data\n",
    "alphas = sorted(set([r['alpha'] for r in dist2run2results['negative-binomial']['gs2_apr_phe']]))\n",
    "alphas.reverse()\n",
    "\n",
    "# manual set which alphas to plot\n",
    "# alphas = [0.005, 0.0025, 0.001, 0.0005]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 151,
   "metadata": {},
   "outputs": [],
   "source": [
    "# pretty pastel-like colors\n",
    "colors = sns.color_palette(\"cubehelix\", n_colors=len(alphas))\n",
    "\n",
    "# manually defined colors\n",
    "# colors = ['#DDCA70', '#235555', '#4198C7', '#464162']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 167,
   "metadata": {},
   "outputs": [],
   "source": [
    "ylimits = {\n",
    "    'negative-binomial': {\n",
    "        'gs2_apr_snp': {\n",
    "            'y': (10e-4, 1.),\n",
    "            'x': (40000, 80000),\n",
    "        },\n",
    "        'gs2_apr_phe': {\n",
    "            'y': (10e-7, 1.),\n",
    "            'x': (40000, 80000),\n",
    "        },\n",
    "        'gs2_may_snp': {\n",
    "            'y': (10e-4, 1.),\n",
    "            'x': (40000, 80000),\n",
    "        },\n",
    "        'gs2_may_phe': {\n",
    "            'y': (10e-7, 1.),\n",
    "            'x': (40000, 80000),\n",
    "        },\n",
    "        'ld_apr_phe': {\n",
    "            'y': (10e-5, 1.),\n",
    "            'x': (0, 20000),\n",
    "        },\n",
    "        'ld_may_phe': {\n",
    "            'y': (10e-5, 1.),\n",
    "            'x': (0, 20000),\n",
    "        }\n",
    "    }\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "for dist_name, run2results in dist2run2results.items():\n",
    "    for run_code, results in run2results.items():\n",
    "\n",
    "        sns.set(style='ticks', font_scale=1.2)\n",
    "        plt.figure(figsize=(6,4))\n",
    "\n",
    "        \n",
    "\n",
    "        for i, a in enumerate(alphas):\n",
    "            results_for_alpha = [(r['distance'], r['nhits_pvalue'], r['chi_pvalue']) for r in results if r['alpha']==a]\n",
    "            \n",
    "            if len(results_for_alpha) == 0:\n",
    "                continue\n",
    "\n",
    "            x, y, gofs = zip(*results_for_alpha)\n",
    "            \n",
    "            pinsig_gofs = len([g for g in gofs if g >= 0.05])/len(gofs)\n",
    "#             plt.plot(x,y, label=r'$\\alpha = %s (%.2f)$' % (a, pinsig_gofs), color=colors[i])\n",
    "            plt.plot(x,y, label=r'$\\alpha = %s$' % a, color=colors[i], lw=3)\n",
    "            plt.yscale('log')\n",
    "            plt.ylabel('Significance of Number of Hits')\n",
    "            plt.xlabel('Distance to Nearest Gene')\n",
    "            \n",
    "            if dist_name in ylimits:\n",
    "                plt.ylim(ylimits[dist_name][run_code]['y'])\n",
    "                plt.xlim(ylimits[dist_name][run_code]['x'])\n",
    "\n",
    "        plt.plot([0, 80000], [0.05, 0.05], 'k--', label=r'$P=0.05%$')\n",
    "        plt.legend(bbox_to_anchor=(1, 1), loc='upper left', ncol=1)\n",
    "        plt.title(run_code + ' ' + dist_name)\n",
    "\n",
    "        sns.despine()\n",
    "        plt.tight_layout()\n",
    "        plt.savefig('figures/r1/permutation-analysis/%s_sighits_%s_minmaf%.3f_0to80kb.pdf' % (run2output_file_prefix[run_code], dist_name, minimum_maf))\n",
    "        \n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "dist_name = 'negative-binomial'\n",
    "run2results = dist2run2results[dist_name]\n",
    "\n",
    "for run_code, results in run2results.items():\n",
    "    \n",
    "    sns.set(style='ticks', font_scale=1.2)\n",
    "    plt.figure(figsize=(6,4))\n",
    "\n",
    "    for i, a in enumerate(alphas):\n",
    "        results_for_alpha = [(r['distance'], r['nhits']) for r in results if r['alpha']==a]\n",
    "        if len(results_for_alpha) == 0:\n",
    "            continue\n",
    "\n",
    "        x, y = zip(*results_for_alpha)\n",
    "        plt.plot(x,y, label=r'$\\alpha = %s$' % a, color=colors[i], lw=3)\n",
    "    #     plt.yscale('log')\n",
    "        plt.ylim(0,45)\n",
    "        plt.ylabel('Number of Hits')\n",
    "        plt.xlabel('Distance to Nearest Gene')\n",
    "\n",
    "    plt.legend(bbox_to_anchor=(1, 1), loc='upper left', ncol=1)\n",
    "    plt.title(run_code)\n",
    "\n",
    "    sns.despine()\n",
    "    plt.tight_layout()\n",
    "    plt.savefig('figures/r1/permutation-analysis/%s_numhits_minmaf%.3f_0to80kb.pdf' % (run2output_file_prefix[run_code], minimum_maf))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 161,
   "metadata": {},
   "outputs": [],
   "source": [
    "alphas = sorted(set([r['alpha'] for r in dist2run2results['negative-binomial']['gs2_apr_phe']]))\n",
    "alphas.reverse()\n",
    "\n",
    "# pretty pastel-like colors\n",
    "colors = sns.color_palette(\"cubehelix\", n_colors=len(alphas))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
